gslib/utils/copy_helper.py

```diff
@@ -663,41 +654,36 @@ def ConstructDstUrl(src_url,
   #    should create objects named like gs://bucket/subdir/dir2/a/b/c. In
   #    contrast, if gs://bucket/subdir does not exist, this same command
   #    should create objects named like gs://bucket/subdir/a/b/c.
-  # 3. Copying individual files or objects to dirs, buckets or bucket subdirs
+  # 2. Copying individual files or objects to dirs, buckets or bucket subdirs
   #    should result in objects/files named by the final source file name
   #    component. Example:
   #      gsutil cp dir1/*.txt gs://bucket
   #    should create the objects gs://bucket/f1.txt and gs://bucket/f2.txt,
   #    assuming dir1 contains f1.txt and f2.txt.
 
-  recursive_move_to_new_subdir = False
-  if (global_copy_helper_opts.perform_mv and recursion_requested and
-      src_url_names_container and not have_existing_dest_subdir):
-    # Case 1. Handle naming rules for bucket subdir mv. Here we want to
-    # line up the src_url against its expansion, to find the base to build
-    # the new name. For example, running the command:
-    #   gsutil mv gs://bucket/abcd gs://bucket/xyz
-    # when processing exp_src_url=gs://bucket/abcd/123
-    # exp_src_url_tail should become /123
-    # Note: mv.py code disallows wildcard specification of source URL.
-    recursive_move_to_new_subdir = True
-    exp_src_url_tail = (exp_src_url.url_string[len(src_url.url_string):])
-    dst_key_name = '%s/%s' % (exp_dst_url.object_name.rstrip('/'),
-                              exp_src_url_tail.strip('/'))
-
-  elif src_url_names_container and (exp_dst_url.IsCloudUrl() or
-                                    exp_dst_url.IsDirectory()):
+  # If there are multiple top-level source items, copy full file hierarchies.
+  # This avoids conflicts such as "dir1/f.txt" and "dir2/f.txt" both getting
+  # copied to "gs://bucket/f.txt". Linux normally errors on these conflicts,
+  # but we cannot do that because we need to give users the ability to create
+  # dirs as they copy to the cloud.
+  # Ignore this rule if using double wildcard ** because that treats all files
+  # as top-level, in which case the user doesn't want to preserve directories.
+  preserve_src_top_level_dirs = '**' not in src_url.versionless_url_string and (
+      has_multiple_top_level_srcs or have_existing_dest_subdir)
+  if preserve_src_top_level_dirs or (src_url_names_container and
```

Comment:
Do we need the case (src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory())) 

I am a bit confused here. Based on the description about for case 1, seems like we are handing only the preserve_src_top_level_dirs for case 1. This means a rename like operation `gsutil mv dir gs://non-existing_dir` will be handled by the else part correct? So do we need to have anything other than preserve_src_top_level_dirs here?

Reply:
I responded to the [earlier thread](https://github.com/GoogleCloudPlatform/gsutil/pull/1465#discussion_r813417927) about this line by explaining the boolean logic differences. Also note there is a `if not preserve_src_top_level_dirs` block for handling the case `(src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory()))`.

And tests fail without the case. Here's one example:
```
FAIL: testCopyingSubdirRecursiveToNonexistentSubdir (gslib.tests.test_naming.GsutilNamingTests)
Tests copying a directory with a single file recursively to a bucket.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/hartunian/Desktop/ide/gsutil/gslib/tests/test_naming.py", line 634, in testCopyingSubdirRecursiveToNonexistentSubdir
    self.assertEqual(expected, actual)
AssertionError: Items in the first set but not the second:
'gs://gsutil-test-testcopyingsubdirrecursivetononexisten-5c92d77f/dir3/dir2/foo'
Items in the second set but not the first:
'gs://gsutil-test-testcopyingsubdirrecursivetononexisten-5c92d77f/dir3/foo'
```

Reply:
Got it. Thanks for checking